core/src/main/java/kafka/log/remote/RemoteLogManager.java

```diff
@@ -1071,15 +1085,38 @@ void cleanupExpiredRemoteLogSegments() throws RemoteStorageException, ExecutionE
                 Iterator<Integer> epochsToClean = remoteLeaderEpochs.stream()
                         .filter(remoteEpoch -> remoteEpoch < earliestEpochEntry.epoch)
                         .iterator();
+
+                List<RemoteLogSegmentMetadata> listOfSegmentsToBeCleaned = new ArrayList<>();
+
                 while (epochsToClean.hasNext()) {
                     int epoch = epochsToClean.next();
                     Iterator<RemoteLogSegmentMetadata> segmentsToBeCleaned = remoteLogMetadataManager.listRemoteLogSegments(topicIdPartition, epoch);
                     while (segmentsToBeCleaned.hasNext()) {
-                        if (isCancelled() || !isLeader()) {
-                            return;
+                        if (!isCancelled() && isLeader()) {
+                            RemoteLogSegmentMetadata nextSegmentMetadata = segmentsToBeCleaned.next();
+                            sizeOfDeletableSegmentsBytes += nextSegmentMetadata.segmentSizeInBytes();
+                            listOfSegmentsToBeCleaned.add(nextSegmentMetadata);
                         }
+                    }
+                }
+
+                segmentsLeftToDelete += listOfSegmentsToBeCleaned.size();
+                brokerTopicMetrics.recordRemoteDeleteLagBytes(partition, sizeOfDeletableSegmentsBytes);
+                brokerTopicMetrics.recordRemoteDeleteLagSegments(partition, segmentsLeftToDelete);
+                for (RemoteLogSegmentMetadata segmentMetadata : listOfSegmentsToBeCleaned) {
+                    if (!isCancelled() && isLeader()) {
                         // No need to update the log-start-offset even though the segment is deleted as these epochs/offsets are earlier to that value.
-                        remoteLogRetentionHandler.deleteLogSegmentsDueToLeaderEpochCacheTruncation(earliestEpochEntry, segmentsToBeCleaned.next());
+                        if (remoteLogRetentionHandler.deleteLogSegmentsDueToLeaderEpochCacheTruncation(earliestEpochEntry, segmentMetadata)) {
+                            sizeOfDeletableSegmentsBytes -= segmentMetadata.segmentSizeInBytes();
+                            segmentsLeftToDelete--;
+                            brokerTopicMetrics.recordRemoteDeleteLagBytes(partition, sizeOfDeletableSegmentsBytes);
+                            brokerTopicMetrics.recordRemoteDeleteLagSegments(partition, segmentsLeftToDelete);
+                        }
+                    } else {
```

Comment:
Why do we need this `else` branch?

Reply:
In my head, we want to stop reporting the lag for partitions which the current replica is no longer the leader for. In other words, there are two conditions when we want to remove the lag from the cumulative - either we have successfully deleted the segment or we are no longer the leader for those partitions. The else condition satisfies the second

Reply:
When the replica becomes follower, we are closing/removing those metrics. If the concern is about concurrency, then I would suggest to mark the `rlmTask` as follower first before removing those metrics. 

In RemoteLogManager:

1. followerPartitions.forEach(topicIdPartition ->
                    doHandleLeaderOrFollowerPartitions(topicIdPartition, RLMTask::convertToFollower));
                    
 2. Then, call followerPartitions.forEach(this::removeRemoteTopicPartitionMetrics);

Reply:
Fair enough, okay, for the current version I will just remove the else-statement and fix the conflicts. I shall have a newer version in a jiffy